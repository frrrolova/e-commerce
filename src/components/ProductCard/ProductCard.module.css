.container {
  position: relative;
  padding-top: 60px;
}
.container .card {
  border-radius: 50px;
  opacity: 0.9;
  height: 280px;
  display: flow-root;
}
.container .card::before {
  content: "";
  position: absolute;
  top: 60px;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 50px;
  border: 2px solid transparent;
  background: linear-gradient(45deg, rgba(38, 46, 34, 0.8), rgba(255, 255, 255, 0.5), rgba(38, 46, 34, 0.8)) border-box;
  -webkit-mask: linear-gradient(#fff 0 0) padding-box, linear-gradient(#fff 0 0);
  -webkit-mask-composite: destination-out;
  mask-composite: exclude;
}
.container .card .content {
  position: absolute;
  top: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  padding-left: 1.5rem;
  padding-right: 1.5rem;
}
.container .card .content p {
  max-height: 3em;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
}
.container .card .image {
  height: 200px;
  transition: transform 500ms ease-in-out;
  filter: drop-shadow(5px 10px 15px rgba(8, 9, 13, 0.4));
}
.container .card:hover .image {
  transform: scale(1.05);
}/*# sourceMappingURL=ProductCard.module.css.map */