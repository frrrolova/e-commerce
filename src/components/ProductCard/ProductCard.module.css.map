{"version":3,"sources":["ProductCard.module.scss","../../styles/variables.scss","ProductCard.module.css"],"names":[],"mappings":"AAEA;EACE,kBAAA;EACA,iBCSmB;ACVrB;AFGE;EACE,mBCOiB;EDNjB,YCOW;EDNX,aCGkB;EDFlB,kBAAA;AEDJ;AFGI;EACE,WAAA;EACA,kBAAA;EACA,SCFe;EDGf,OAAA;EACA,QAAA;EACA,SAAA;EACA,mBCLe;EDMf,6BAAA;EACA,qHAAA;EACA,8EAAA;EACA,uCAAA;EACA,uBAAA;AEDN;AFII;EACE,kBAAA;EACA,MAAA;EACA,aAAA;EACA,sBAAA;EACA,mBAAA;EACA,WAAA;EACA,oBAAA;EACA,qBAAA;AEFN;AFIM;EACE,eAAA;EACA,gBAAA;EACA,uBAAA;EACA,oBAAA;EACA,4BAAA;EACA,qBAAA;AEFR;AFMI;EACE,aCvCiB;EDwCjB,uCAAA;EACA,sDAAA;AEJN;AFQM;EACE,sBAAA;AENR","file":"ProductCard.module.css"}