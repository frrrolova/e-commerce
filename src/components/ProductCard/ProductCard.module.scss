@import '../../styles/variables.scss';

.container {
  position: relative;
  padding-top: $offset-product-top;

  .card {
    border-radius: $border-radius-card;
    opacity: $opacity-item;
    height: $height-product-card;
    display: flow-root;

    &::before {
      content: "";
      position: absolute;
      top: $offset-product-top;
      left: 0;
      right: 0;
      bottom: 0;
      border-radius: $border-radius-card;
      border: 2px solid transparent;
      background: linear-gradient(45deg, $border-dark-color, $border-light-color, $border-dark-color) border-box;
      -webkit-mask: linear-gradient(#fff 0 0) padding-box, linear-gradient(#fff 0 0);
      -webkit-mask-composite: destination-out;
      mask-composite: exclude;
    }

    .content {
      position: absolute;
      top: 0;
      display: flex;
      flex-direction: column;
      align-items: center;
      width: 100%;
      padding-left: 1.5rem;
      padding-right: 1.5rem;

      p {
        max-height: 3em;
        overflow: hidden;
        text-overflow: ellipsis;
        display: -webkit-box;
        -webkit-box-orient: vertical;
        -webkit-line-clamp: 2;
      }
    }

    .image {
      height: $height-product-image;
      transition: transform 500ms ease-in-out;
      filter: drop-shadow(5px 10px 15px $shadow-color);
    }

    &:hover {
      cursor: pointer;
      background-color: #323d2d;
      transition: all 500ms ease-in-out;
      .image {
        transform: scale(1.05);
      }
    }

    .description {
      min-width: 100%;
    }

    .bottomContent {
      display: flex;
      align-items: center;

      .discount {
        font-size: larger;
        font-weight: bold;
        color: $border-green-color;
        border-right: 2px solid $border-light-color;
      }
    }
  }
}
