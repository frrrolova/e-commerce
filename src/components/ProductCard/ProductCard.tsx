import styles from './ProductCard.module.scss';
import { Box, Paper, Typography } from '@mui/material';
import { Product } from '@/types';

interface ProductCardProps {
  product: Product;
}

function ProductCard({ product }: ProductCardProps) {
  return (
    <Box className={styles.container}>
      <Paper elevation={3} className={styles.card}>
        <Box className={styles.content}>
          <Box component="img" className={styles.image} alt="Plant" src={product.imgPath} />
          <Box className={styles.description}>
            <Typography gutterBottom variant="h5" component="div" mt={1}>
              {product.label}
            </Typography>
            <Typography variant="body2" color="text.secondary">
              {product.description}
            </Typography>
            {/* TODO Add redirect to Product page*/}
            {/* <Button size="small" sx={{ mt: 1 }}>
              Buy Now
            </Button> */}
          </Box>
        </Box>
      </Paper>
    </Box>
  );
}

export default ProductCard;
