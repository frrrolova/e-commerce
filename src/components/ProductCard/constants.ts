export const centsInEuro = 100;

export enum PageData {
  addTooltip = 'Add to Cart',
  addSnackbar = 'Product was added to cart',
}
