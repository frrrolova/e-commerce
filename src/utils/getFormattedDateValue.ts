export function getFormattedDateValue(value: string) {
  const date = new Date(value);
  const dateValue: string = new Date(date.getTime() - date.getTimezoneOffset() * 60000).toISOString().substring(0, 10);
  return dateValue;
}
