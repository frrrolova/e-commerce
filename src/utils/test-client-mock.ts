import { Category, Product, PromoData } from '@/types';

export const mockProduct1: Product = {
  id: '1',
  name: 'Product Test1',
  description: 'Description 1',
  images: [{ url: 'image.png', dimensions: { w: 200, h: 200 } }],
  prices: [{ id: '1', value: { type: 'centPrecision', centAmount: 2000, currencyCode: 'EUR', fractionDigits: 2 } }],
};

export const mockProduct2: Product = {
  id: '2',
  name: 'Product Test2',
  description: 'Description 2',
  images: [{ url: 'image2.png', dimensions: { w: 200, h: 200 } }],
  prices: [{ id: '2', value: { type: 'centPrecision', centAmount: 2000, currencyCode: 'EUR', fractionDigits: 2 } }],
};

export const mockCategoryParent: Category = {
  id: '3',
  name: 'Category Parent',
  key: 'category-parent-key',
  ancestors: [],
  parent: undefined,
};

export const mockCategoryChild: Category = {
  id: '4',
  name: 'Category Child',
  key: 'category-child-key',
  ancestors: [{ typeId: 'category', id: '3' }],
  parent: { typeId: 'category', id: '3' },
};

export const mockPromoData: PromoData = {
  id: '123',
  heading: 'Promo Name',
  imgPath: '/img',
  description: 'promo description',
  subHeading: 'promocode',
};

export const clientMock = () => ({
  getClient: () => ({
    productProjections: () => ({
      search: () => ({
        get: jest.fn().mockReturnValue({
          execute: jest.fn().mockResolvedValue({ body: { results: [mockProduct2] } }),
        }),
      }),
      withId: () => ({
        get: jest.fn().mockReturnValue({
          execute: jest.fn().mockResolvedValue({ body: mockProduct1 }),
        }),
      }),
      get: jest.fn().mockReturnValue({
        execute: jest.fn().mockResolvedValue({ body: { results: [mockProduct2] } }),
      }),
    }),
    productTypes: () => ({
      get: jest.fn().mockReturnValue({
        execute: jest.fn().mockResolvedValue({ body: { results: [] } }),
      }),
    }),
    categories: () => ({
      get: jest.fn().mockReturnValue({
        execute: jest.fn().mockResolvedValue({ body: { results: [mockCategoryParent, mockCategoryChild] } }),
      }),
    }),
    cartDiscounts: () => ({
      get: jest.fn().mockReturnValue({
        execute: jest.fn().mockResolvedValue({ body: { results: [mockPromoData] } }),
      }),
    }),
  }),
});
