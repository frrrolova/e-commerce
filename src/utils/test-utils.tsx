import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';

import type { AppStore, RootState } from '../store/store';
// As a basic setup, import your same slice reducers
import initStore from '../store/store';
import { MemoryRouter } from 'react-router-dom';
import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { Project } from '@commercetools/platform-sdk';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

const mockShop: Project = {
  countries: [],
  createdAt: new Date().toISOString(),
  languages: [],
  key: 'test-key',
  version: 2,
  name: 'Plant Shop',
  currencies: [],
  messages: { enabled: false, deleteDaysAfterCreation: 15 },
  carts: {
    deleteDaysAfterLastModification: 90,
    countryTaxRateFallbackEnabled: false,
  },
};

export function renderWithProviders(ui: React.ReactElement, extendedRenderOptions: ExtendedRenderOptions = {}) {
  const {
    preloadedState = {
      project: mockShop,
    },
    // Automatically create a store instance if no store was passed in
    store = initStore(preloadedState.project as Project),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <MemoryRouter>
      <LocalizationProvider dateAdapter={AdapterDayjs}>
        <Provider store={store}>{children}</Provider>
      </LocalizationProvider>
    </MemoryRouter>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
