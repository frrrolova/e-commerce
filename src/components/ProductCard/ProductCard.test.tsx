import ProductCard from './ProductCard';
import { waitFor, screen, fireEvent } from '@testing-library/react';
import { act } from 'react';
import { clientMock, mockProduct1 } from '@/utils/test-client-mock';
import { renderWithProviders } from '@/utils/test-utils';

// Mock for client
jest.mock('../../client/client', () => {
  return {
    client: jest.fn().mockImplementation(() => clientMock),
  };
});

//Mock for useNavigate
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
  useParams: jest.fn().mockReturnValue({ productId: 'testId' }),
}));

jest.mock('@/services/productService', function () {
  return {
    productService: {
      fetchProduct: jest.fn().mockResolvedValue({
        id: 'testId',
        name: 'testName',
        description: 'testDescription',
        images: [
          {
            url: 'testUrl',
            label: 'testLabel',
          },
        ],
        prices: [
          {
            value: {
              centAmount: 1000,
            },
            discounted: {
              value: {
                centAmount: 500,
              },
            },
          },
        ],
      }),
    },
  };
});

// Mock for notistack
jest.mock('notistack', () => ({
  enqueueSnackbar: jest.fn(),
}));

describe('Main component rendering', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('performs snapshot testing', async () => {
    let tree;
    await act(async () => {
      tree = renderWithProviders(<ProductCard product={mockProduct1} />);
    });

    await waitFor(() => {
      const cardElement = screen.getByText(mockProduct1.name);
      expect(cardElement).toBeInTheDocument();
    });

    expect(tree).toMatchSnapshot();
  });
});

describe('Card component behavior', () => {
  test('should navigate, when clicking on a card', async () => {
    const { getByText } = renderWithProviders(<ProductCard product={mockProduct1} />);

    await waitFor(() => {
      const button = getByText(mockProduct1.name);

      fireEvent.click(button);
      expect(mockNavigate).toHaveBeenCalledTimes(1);
    });
  });
});
