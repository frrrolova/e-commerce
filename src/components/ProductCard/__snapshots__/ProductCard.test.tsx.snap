// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component rendering performs snapshot testing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <img
              alt="Plant"
              class="MuiBox-root css-0"
              loading="lazy"
              src="image.png"
            />
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-ix1vfo-MuiTypography-root"
              >
                Product Test1
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                Description 1
              </p>
            </div>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-0"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                20 €
              </p>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <span
                aria-label="Add to Cart"
                class=""
                data-mui-internal-clone-element="true"
              >
                <button
                  aria-label="add"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-16rd89w-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="AddShoppingCartIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2m10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2m-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <img
            alt="Plant"
            class="MuiBox-root css-0"
            loading="lazy"
            src="image.png"
          />
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-ix1vfo-MuiTypography-root"
            >
              Product Test1
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              Description 1
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              20 €
            </p>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <span
              aria-label="Add to Cart"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                aria-label="add"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-16rd89w-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddShoppingCartIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2m10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2m-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
